packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,89 @@
+/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0 */
+import { IRole, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
+import { IBucket } from "aws-cdk-lib/aws-s3";
+import {
+  ApiGatewayIntegration,
+  Integration,
+  IntegrationRenderProps,
+} from "./integration";
+import { generateCorsResponseParameters } from "../prepare-spec-event-handler/prepare-spec";
+import { bucketInvocationUri } from "../spec/utils";
+
+export interface S3IntegrationProps {
+  /**
+   * The S3 bucket to be invoked on integration
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * The IAM role to be used to grant permissions to the API Gateway to invoke the S3 bucket
+   * @default - a new role will be created
+   */
+  readonly role?: IRole;
+
+  /**
+   * The path override for the integration
+   * @default - integration path is used
+   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html
+   */
+  readonly pathOverride?: string;
+}
+
+/**
+ * An S3 integration
+ */
+export class S3Integration extends Integration {
+  private readonly bucket: IBucket;
+  private readonly role: IRole;
+  private readonly pathOverride?: string;
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+
+    if (props.role) {
+      this.role = props.role;
+    } else {
+      this.role = new Role(this.bucket, "Role", {
+        assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
+      });
+    }
+
+    this.bucket.grantReadWrite(this.role, this.pathOverride ?? "*");
+    this.pathOverride = props.pathOverride;
+  }
+
+  /**
+   * Render the lambda integration as a snippet of OpenAPI
+   */
+  public render(props: IntegrationRenderProps): ApiGatewayIntegration {
+    return {
+      type: "AWS",
+      httpMethod: props.method.toUpperCase(),
+      uri: bucketInvocationUri(
+        props.scope,
+        this.bucket,
+        this.pathOverride ?? props.path
+      ),
+      credentials: this.role.roleArn,
+      responses: {
+        default: {
+          statusCode: "200",
```

Comment:
That's the tricky part because S3 can send a lot of errors.

Maybe a way to handle it is to provide different implementations:
- one that do nothing
- one that simply aggregate all 4XX and all 5XX to a simple error
- one that totally customizable by the user

Reply:
Nice idea! Let's go with that approach if it's not too hard to build :) If the totally customizable one is too difficult to implement we can skip it, as users can always define a custom integration themselves :)

Reply:
@cogwirrel what do you think of the proposed `ErrorIntegrationResponse`?

Reply:
Great, and it could be used in other integrations too, but it's totally useless in LambdaIntegration because API Gateway act as a proxy and that's the lambda job to return errors